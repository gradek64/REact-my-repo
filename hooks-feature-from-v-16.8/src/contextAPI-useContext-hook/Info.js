import React, { useContext } from 'react';
// custom
import { SectionContext } from './index';

const Info = () => {
  const { list } = useContext(SectionContext);
  return `curtent number of items: ${list.items.length}`;
};

export default Info;
