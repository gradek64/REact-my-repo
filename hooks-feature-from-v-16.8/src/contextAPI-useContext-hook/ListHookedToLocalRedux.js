import React, { useContext } from 'react';
// custom;
import { SectionContext } from './index';

const List = () => {
  const { list, dispatchAction } = useContext(SectionContext);

  return (
    <div>
      {list.items.map(({ title }, i) => (
        <div key={`keys${i}`}>
          <span>title</span>
          <button
            type="button"
            onClick={() => dispatchAction({ type: 'REMOVE_ITEM', title })}
          >
            remove
          </button>
        </div>
      ))}
    </div>
  );
};
export default List;
