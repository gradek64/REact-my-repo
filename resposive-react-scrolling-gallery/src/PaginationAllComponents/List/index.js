import React from 'react';
import PropTypes from 'prop-types';

// custom
import Card from '../Item/Card';

/* eslint-disable no-underscore-dangle */
const List = ({ list }) => (
  <div className="paginated-list">
    <ul className="cards-slider-inner">
      {
        list.map((item, key) => <Card key={item._id} {...item} />)
      }
    </ul>
  </div>
);

List.propTypes = {
  list: PropTypes.array.isRequired,
};

export default List;
