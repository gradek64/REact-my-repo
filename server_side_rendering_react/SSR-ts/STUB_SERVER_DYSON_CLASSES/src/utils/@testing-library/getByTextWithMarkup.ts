import { screen } from '@testing-library/react'

const matcher = (text) => {
  return (_, node: HTMLElement) => {
    const hasText = (innerNode: HTMLElement) => innerNode.textContent === text
    const childrenDontHaveText = Array.from(node.children).every((child: HTMLElement) => !hasText(child))
    return hasText(node) && childrenDontHaveText
  }
}

export const getByTextWithMarkup = (text: string): ReturnType<(typeof screen)['getByText']> =>
  screen.getByText(matcher(text))

export const queryByTextWithMarkup = (text: string): ReturnType<(typeof screen)['queryByText']> =>
  screen.queryByText(matcher(text))
