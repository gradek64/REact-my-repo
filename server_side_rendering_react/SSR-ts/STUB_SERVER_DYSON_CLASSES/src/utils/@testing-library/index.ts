export * from './renderer'
export { waitForAnimatedSubmitButton } from './waitForAnimatedSubmitButton'
export { getByTextWithMarkup, queryByTextWithMarkup } from './getByTextWithMarkup'

// eslint-disable-next-line import/no-extraneous-dependencies
export { default as userEvent } from '@testing-library/user-event'

export { default as creditPlanSelectors } from './selectors/creditPlans'
export { renderWithModalWrapper } from './renderWithModalWrapper'
