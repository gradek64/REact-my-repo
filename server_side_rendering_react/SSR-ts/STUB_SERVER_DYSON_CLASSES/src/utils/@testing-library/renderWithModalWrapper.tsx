import React, { ReactNode } from 'react'
import { Modal } from '@sainsburys-tech/bolt'
import { RenderResult, render } from '.'

export const renderWithModalWrapper = (children: ReactNode, renderOptions = {}, modalProps = {}): RenderResult =>
  render(
    <Modal show {...modalProps}>
      {children}
    </Modal>,
    renderOptions,
  )
