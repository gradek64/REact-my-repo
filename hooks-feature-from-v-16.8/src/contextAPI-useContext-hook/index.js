import React, { useReducer, useContext } from 'react';
import List from './ListHookedToLocalRedux';
import { listReducer } from './listReducer';
import Info from './Info';

/*
      Context provides a way to pass data through the component tree

      without having to pass props down manually at every level

      --React quatation

      1. create context
      2. <Context>...</Context> wrappet text a value as prop with initial object to be shared
      2a. It is common use to give initial context value from useReducer() hook

*/
//needs to export for testing and accesiblity reason;
export const initialList = {
  items: [
    { title: 'title1' },
    { title: 'title2' },
    { title: 'title3' },
    { title: 'title4' },
    { title: 'title5' },
  ],
  greg: 'is so cool',
};

// create context that would wrap all children components ang give access to context valuse
// it needs to be exported since other component have to have acccess to it by importing it
export const SectionContext = React.createContext();
const { Provider } = SectionContext;

const AppLocalRedux = () => {
  //this is same as useState but with state management as redux will do
  const [list, dispatchAction] = useReducer(listReducer, initialList);

  return (
    <div style={{ border: '1px dashed blue' }}>
      {/* we have wrapper that provides state distribution in a value prop */}
      <Provider value={{ list, dispatchAction }}>
        {/*Notice bellow components dont have props they receive data by context wrapper*/}
        {/* props are encapsulated in value object which has dispatchAction method and list*/}
        <Info />
        <List />
      </Provider>
    </div>
  );
};

export default AppLocalRedux;
