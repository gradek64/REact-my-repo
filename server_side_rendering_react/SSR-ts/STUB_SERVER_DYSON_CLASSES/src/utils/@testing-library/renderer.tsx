/* eslint-disable @typescript-eslint/no-unsafe-member-access */
/* eslint-disable @typescript-eslint/no-unsafe-assignment */
import React, { FC, ReactElement } from 'react'
// eslint-disable-next-line no-restricted-imports
import { render, configure, RenderResult } from '@testing-library/react'

import { Provider } from 'react-redux'
import { createBrowserHistory } from 'history'
import { ConnectedRouter } from 'connected-react-router'
import { ThemeProvider } from 'styled-components'
import { Themes } from '@sainsburys-tech/bolt'
import { I18nextProvider } from 'react-i18next'
import { PreloadedState } from 'redux'

import i18n from '../../../config/i18n/test'
import rootSaga from '../../sagas'
import createStoreActual from '../../store'
import { RootState } from '../../types/reducers'
import { initialState as reducerInitialState } from '../../config/testUtils/__mocks__/state'

const browserHistory = createBrowserHistory()

configure({ testIdAttribute: 'data-test' })

export const createStore = (
  initialState: PreloadedState<RootState>,
  initialHistory: typeof browserHistory | undefined,
): ReturnType<typeof createStoreActual> => {
  const history = initialHistory || browserHistory
  const store = createStoreActual(history, initialState)

  store.runSaga(rootSaga)
  return store
}

/* https://testing-library.com/docs/react-testing-library/setup */
const customRender = (
  ui: ReactElement,
  {
    initialState = reducerInitialState,
    initialHistory = browserHistory,
    store = createStore(initialState, initialHistory),
    ...renderOptions
  } = {},
): RenderResult => {
  const AllTheProviders: FC = ({ children }) => {
    return (
      <I18nextProvider i18n={i18n}>
        <Provider store={store}>
          <ConnectedRouter history={initialHistory}>
            <ThemeProvider theme={Themes.ArgosTheme}>{children}</ThemeProvider>
          </ConnectedRouter>
        </Provider>
      </I18nextProvider>
    )
  }
  return render(ui, { wrapper: AllTheProviders, ...renderOptions })
}

// eslint-disable-next-line no-restricted-imports
export * from '@testing-library/react'

export { customRender as render }
