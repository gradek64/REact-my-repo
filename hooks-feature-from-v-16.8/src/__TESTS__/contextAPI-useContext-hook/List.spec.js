import React from 'react';
import { render, cleanup, fireEvent } from '@testing-library/react';
import '@testing-library/jest-dom/extend-expect';
import ContextAPI from '../../contextAPI-useContext-hook';

//import child testing component
import List from '../../contextAPI-useContext-hook/ListHookedToLocalRedux';
//import data used in context < <Context.Provider>
//import { initialList } from '../../contextAPI-useContext-hook/index';
import ListParent, {
  initialList,
} from '../../contextAPI-useContext-hook/index';
//import context
import { SectionContext } from '../../contextAPI-useContext-hook/index';

//=========================================================
// in order to test useContext() hook you need to provide
// context that comes from parent component
// also data  has to be the same as used in parent component
//=========================================================

/**
 * order of action:
 *
 * @1 import testing child component <List />
 * @2 import data used in parent component {initialList}
 * @3 import context from React  export const SectionContext = React.createContext();
 * @4 wrap child component into Context
 * @5 skip parent from rendering since if you do it will render all children
 *    from context, but we only intrested in <Info /> child component
 * @6 to fail a test change data in original component what will affect snaposhot
 *    for <Info> component
 */
const { items } = initialList;
describe('with mock Provider', () => {
  let ComponentWithContext;
  let dispatchAction = jest.fn();
  beforeEach(() => {
    ComponentWithContext = () => (
      //<InfoParent> parent not need since it will render all children
      //not just <List /> tested component
      <SectionContext.Provider value={{ list: { items }, dispatchAction }}>
        <List />
      </SectionContext.Provider>
      // </InfoParent>
    );
  });
  //important to clean afger every test!
  afterEach(cleanup);
  describe('snapshot for <List Component', () => {
    it('it renders List component correctlly', () => {
      const { asFragment } = render(<ComponentWithContext />);
      expect(asFragment(<ComponentWithContext />)).toMatchSnapshot();
    });
  });

  describe('mock events', () => {
    it('calls dispatchAction on first button click', () => {
      /**
       *
       * @usage: going by the first apprach 'mock provider'
       * we can mock dispach method given from context and
       * chack if our method has been fired with function argms
       * set in original imlementation as below
       **/

      const { container } = render(<ComponentWithContext />);
      const bttn1 = container.querySelectorAll('button')[0];
      fireEvent.click(bttn1);
      expect(dispatchAction).toHaveBeenCalledWith({
        type: 'REMOVE_ITEM',
        title: items[0].title,
      });
    });
  });
});

/**
 *
 * @usage: another approuch is to not mock it at all and just parentContainer
 * which keeps all context and just does his job just like in application
 * not provider needs test will use just functionallity of <Context.Provider>
 * form parent Component
 *
 *
 **/

describe('events', () => {
  it('Context value is updated by child component', () => {
    /**
     *
     * we could just test if something has been called but
     * I think is better to check if any changes to content change
     * after click not just check if something has been called ?
     *
     *
     **/
    let { container } = render(<ListParent />);
    let allBttns = container.querySelectorAll('button').length;
    const listLength = items.length;
    expect(allBttns).toEqual(listLength);

    //click on first button
    const bttn1 = container.querySelectorAll('button')[0];
    fireEvent.click(bttn1);
    ({ container } = render(<ListParent />));
    expect(container.querySelectorAll('button').length).toEqual(listLength - 1);

    //click on first button after romovale of first one iniatially
    //needs re-render
    const next1 = container.querySelectorAll('button')[0];
    fireEvent.click(next1);
    ({ container } = render(<ListParent />));
    expect(container.querySelectorAll('button').length).toEqual(listLength - 2);
  });
});
