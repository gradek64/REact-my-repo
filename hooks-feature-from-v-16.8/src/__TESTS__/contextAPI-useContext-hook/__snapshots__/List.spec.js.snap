// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`with mock Provider snapshot for <List Component it renders List component correctlly 1`] = `
<DocumentFragment>
  <div>
    <div>
      <span>
        title
      </span>
      <button
        type="button"
      >
        remove
      </button>
    </div>
    <div>
      <span>
        title
      </span>
      <button
        type="button"
      >
        remove
      </button>
    </div>
    <div>
      <span>
        title
      </span>
      <button
        type="button"
      >
        remove
      </button>
    </div>
    <div>
      <span>
        title
      </span>
      <button
        type="button"
      >
        remove
      </button>
    </div>
    <div>
      <span>
        title
      </span>
      <button
        type="button"
      >
        remove
      </button>
    </div>
  </div>
</DocumentFragment>
`;
